### PR TITLE
Update finetune_adapter.py

### PR DESCRIPTION
**Pull request by awaelchli**

### CODE REVIEW

File: finetune_adapter.py [hunk 1]
```diff
@@ -34,24 +35,16 @@
...
-    },
+    "zero_optimization": {"stage": 2},
 }
 
 
 def main(
-    data_dir: Path = Path("data/alpaca"), checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b")
+    data_dir: Path = Path("data/alpaca"),
+    pretrained_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b"),
+    out_dir: Path = Path("out/adapter/alpaca"),
```
Comment (awaelchli) [inline finetune_adapter.py:45]: In #31, for some reason last minute commits would merge these two directories. Why? I'm pretty sure we want to separate the input directory and output directory.

File: generate_adapter.py [hunk 2]
```diff
@@ -18,7 +18,8 @@
 def main(
     prompt: str = "What food do lamas eat?",
     input: str = "",
-    checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b"),
+    adapter_path: Optional[Path] = None,
+    pretrained_dir: Optional[Path] = None,
```
Comment (awaelchli) [inline generate_adapter.py:22]: Same here, see comment above. 
For evaluation purposes, it makes sense to be able to read different finetuned checkpoints from a separate directory.

File: finetune_adapter.py [hunk 3]
```diff
@@ -154,7 +151,7 @@ def validate(fabric: L.Fabric, model: torch.nn.Module, val_data: np.ndarray, tok
     fabric.print(instruction)
     sample = {"instruction": instruction, "input": ""}
     prompt = generate_prompt(sample)
-    encoded = tokenizer.encode(prompt, bos=True, device=model.device)
```
Comment (awaelchli) [inline finetune_adapter.py:157]: The bos token doesn't seem to be necessary. The loss convergence seems to be unaffected though.

File: finetune_adapter.py [hunk 4]
```diff
@@ -34,24 +35,16 @@
...
-    "zero_optimization": {
-        "stage": 2,
-        # "overlap_comm": True,
-        # "allgather_partitions":True,
-        # "allgather_bucket_size": 2e8,
-        # "reduce_scatter":True,
-        # "contiguous_gradients": True,
-        # "round_robin_gradients": True,
-        "offload_optimizer": {"device": "cpu", "pin_memory": True},
-        "offload_param": {"device": "cpu"},
```
Comment (awaelchli) [inline finetune_adapter.py:46]: offloading would require us to change the optimizer. For the 3B model with 500k trainable parameters for finetuning, this is probably not worth it.

File: generate_adapter.py (hunk 2 above)
Comment (lantiga) [reply generate_adapter.py:22]: 100% agreed

File: finetune_adapter.py [hunk 5]
```diff
@@ -34,24 +35,16 @@
...
-    },
+    "zero_optimization": {"stage": 2},
 }
 
 
 def main(
-    data_dir: Path = Path("data/alpaca"), checkpoint_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b")
+    data_dir: Path = Path("data/alpaca"),
+    pretrained_dir: Path = Path("checkpoints/stabilityai/stablelm-base-alpha-3b"),
```
Comment (carmocca) [inline finetune_adapter.py:44]: Can we keep the `checkpoint_dir` name?

File: finetune_adapter.py (hunk 3 above)
Comment (carmocca) [reply finetune_adapter.py:157]: Can we remove the `bos` and `eos` tokenizer logic to simplify? AFAIK it's now unused

File: finetune_adapter.py [hunk 6]
```diff
@@ -90,15 +86,16 @@ def train(
     optimizer: torch.optim.Optimizer,
     train_data: np.ndarray,
     val_data: np.ndarray,
-    checkpoint_dir: Path,
+    pretrained_dir: Path,
+    out_dir: Path,
```
Comment (carmocca) [inline finetune_adapter.py:90]: Can you update the https://github.com/Lightning-AI/lit-stablelm/blob/main/howto/finetune_adapter.md file to show out_dir usage?

If we don't have a single checkpoint, then we should recommend out_dir to have some sort of information of which checkpoint was used, like `out/adapter/stabilityai.../alpaca`

Alternatively, out_dir could be created inside `checkpoint_dir` and then it would be clear from the structure

File: finetune_adapter.py (hunk 3 above)
Comment (awaelchli) [reply finetune_adapter.py:157]: It is unclear how we would justify removing that. Maybe the bos yes if we are sure that it won't be used. But we should probably do that in a follow up not here, because here I'm just fixing basic stuff for the existing finetuning script.

Comment (carmocca) [reply finetune_adapter.py:157]: I'm okay with doing this separately.

File: finetune_adapter.py (hunk 5 above)
Comment (awaelchli) [reply finetune_adapter.py:44]: In Lit-LLaMA, it is also pretrained_path and this is to disambiguate with the checkpoint directory where we save training checkpoints.

Comment (carmocca) [reply finetune_adapter.py:44]: It is easier to see what this argument expectes if it matches the name in `convert_hf_checkpoint`, `generate`, and other scripts.

Particularly for users who are not knowledgeable about the differences between pertaining, training, fine-tuning etc

File: finetune_adapter.py (hunk 6 above)
Comment (carmocca) [reply finetune_adapter.py:90]: The single folder motivation stems from this review comment: https://github.com/Lightning-AI/lit-stablelm/pull/19#discussion_r1186457956 where it was suggested that we could use a single directory and infer the structure from it.

I understand the motivation for keeping a separate `out_dir` argument, but it could be created inside the "main" directory to be in line with the suggestion above.

Comment (lantiga) [reply finetune_adapter.py:90]: I agree, let's avoid having the finetuned output in the original checkpoint directory.

File: finetune_adapter.py (hunk 5 above)
Comment (awaelchli) [reply finetune_adapter.py]: I reverted back to the ambiguous name but I disagree on every level.

File: finetune_adapter.py [hunk 7]
```diff
@@ -78,10 +74,10 @@ def main(
 
     optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate, weight_decay=weight_decay)
     model, optimizer = fabric.setup(model, optimizer)
-    train(fabric, model, optimizer, train_data, val_data, checkpoint_dir)
+    train(fabric, model, optimizer, train_data, val_data, checkpoint_dir, out_dir)
 
     # Save the final checkpoint at the end of training
-    save_model_checkpoint(fabric, model, checkpoint_dir / "lit_model_adapter_finetuned.pth")
```
Comment (carmocca) [inline finetune_adapter.py:84]: nit: all other files use underscore instead of dashes (`lit_model` , `lit_config`, `tokenizer_config`). The adapter checkpoint should too

File: generate_adapter.py [hunk 8]
```diff
@@ -40,9 +43,12 @@ def main(
         temperature: A value controlling the randomness of the sampling process. Higher values result in more random
             samples.
     """
+    if not adapter_path:
+        adapter_path = Path("out/adapter/alpaca/lit-stablelm-adapter-finetuned.pth")
+    if not checkpoint_dir:
+        checkpoint_dir = Path(f"checkpoints/stabilityai/stablelm-base-alpha-3b")
```
Comment (carmocca) [inline generate_adapter.py:49]: Why are these made optional if we are setting the paths here? This can be the default in the signature

File: finetune_adapter.py (hunk 5 above)
Comment (lantiga) [reply finetune_adapter.py]: I think we should be consistent with lit-llama actually, so I'd personally go with `pretrained_dir` (for the same reason why we chose `pretrained_dir` there)
let's just write a meaningful docstring for that argument so it's clear it coincides with the checkpoint

**Review by carmocca — APPROVED** [review]
Meant to approve ^

File: finetune_adapter.py (hunk 7 above)
Comment (awaelchli) [reply finetune_adapter.py:84]: That's not true, I was always consistent and we have all checkpoints in lit-llama like this. For example see finetune_adapter.py in lit-llama, the same naming there. This does not matter to the practitioner and I'm not going to manually rename all the files we already have just to make this consistent.

File: generate_adapter.py (hunk 8 above)
Comment (awaelchli) [reply generate_adapter.py:49]: It's the same in lit-llama, what's the problem?

Comment (carmocca) [reply generate_adapter.py:49]: It's an improvement suggestion. For example, if this was moved to the signature, the help message would show the true default value.

I can propose the same change in lit-llama if you want it to be the same
